[13.0][IMP] shopfloor: location content transfer

sebalix opened the pull request.

Commit:
location content transfer: add skeleton

Commit:
location transfer: implement /start_or_recover, /scan_location

Commit:
checkout: fix state change when picking not found

Commit:
location transfer: implement /set_destination_all

Commit:
location transfer: implement /go_to_single

Commit:
location transfer: implement /scan_package

Commit:
location transfer: implement /scan_line

Commit:
location transfer: Add a bit of docstring for /scan_location

Commit:
Add todo

sebalix committed:
location transfer: improve /scan_location by creating moves if needed

sebalix committed:
location transfer: add a test for /scan_package if location doesn't exist

sebalix committed:
location transfer: improve /go_to_single to redirect on 'start' screen if there is no move lines to process

sebalix committed:
location transfer: implement /set_destination_package and /set_destination_line endpoints

sebalix committed:
location transfer: implement /postpone_package and /postpone_line endpoints

sebalix committed:
location transfer: implement /stock_out_package and /stock_out_line endpoints

sebalix commented:
- [x] <s>TODO: rebase to include this: https://github.com/camptocamp/wms/pull/29#pullrequestreview-443802030 and update the sort_key for location transfer accordingly.</s>

sebalix committed:
location transfer: handle 'shopfloor_picking_sequence' sort key

guewen commented on shopfloor/actions/data.py:
```diff
@@ -148,6 +148,31 @@ def _move_line_parser(self):
+                    # cannot use sub-parser here
+                    # because location_id of the package level may be
+                    # empty, we have to go get the picking's one
+                    "location_src": self.location(record.picking_id.location_id, **kw),
```
Simone told me that we can use a function/lambda in the subparser, so we could but in `_package_level_parser` directly. (I know I did this but as you are here :stuck_out_tongue_winking_eye:)

simahawk replied:
I've done this here among other fix/imp https://github.com/sebalix/wms/pull/1

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,922 @@
+    def _data_content_all_for_location(self, pickings):
+        if not pickings:
+            # TODO get pickings from location
+            raise NotImplementedError("to do: get pickings from location")
```
isn't it done by the caller of `_data_content_all_for_location` already? Since it's my comment, it may be obsolete now.

sebalix replied:
Doesn't seem, it's called by `_response_for_scan_destination_all`, which is called here: https://github.com/camptocamp/wms/pull/27/files#diff-7105f337bd0a81fafe03414dc6347bdfR334-R347
But `pickings` could be empty there. So I should get pickings from the source location of the picking type related to the menu?

guewen replied:
`_response_for_scan_destination_all` is called by `_router_single_or_all_destination`, the callers of this method already did either search the pickings in the location or provide the pickings. IMO, `_router_single_or_all_destination` should go to the initial screen when it receives no picking. Which also means `start_or_recover()`  can be simplified, it only has to call `_router_single_or_all_destination`

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,922 @@
+                backorder_move = self.env["stock.move"].browse(backorder_move_id)
+                backorder_move.move_line_ids = other_move_lines
+                backorder_move._action_assign()
+        if package_move_lines == package_moves.mapped("move_line_ids"):
```
Why do you check this, isn't it always true?

sebalix replied:
It was for the case where one move uses 1 package and 1 another move.line to be satisfied, so that the move lines related to the package has to be the same than the move lines of the moves related to the package.
Obviously the code above should ensure that we have 1 move for the current package, so yes it's should be always true... Yup better to remove it.

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,922 @@
+
+        NOTE for implementation: Use the field "shopfloor_postponed", which has
+        to be included in the sort to get the next lines.
+
```
```suggestion
```

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,922 @@
+
+        NOTE for implementation: Use the field "shopfloor_postponed", which has
+        to be included in the sort to get the next lines.
+
```
```suggestion
```

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,922 @@
+            # Check if there is no other lines linked to the move others than
+            # the lines related to the package itself. In such case we have to
+            # split the move to process only the lines related to the package.
+            other_move_lines = package_move.move_line_ids - package_move_lines
+            if other_move_lines:
+                qty_to_split = sum(other_move_lines.mapped("product_uom_qty"))
+                backorder_move_id = package_move._split(qty_to_split)
+                backorder_move = self.env["stock.move"].browse(backorder_move_id)
+                backorder_move.move_line_ids = other_move_lines
+                backorder_move._action_assign()
```
maybe this can be extracted in a method to reuse in `set_destination_package`?

sebalix replied:
indeed!

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,922 @@
+            backorder_move_id = move_line.move_id._split(qty_to_split)
+            backorder_move = self.env["stock.move"].browse(backorder_move_id)
+            backorder_move.move_line_ids = other_move_lines
+            backorder_move._action_assign()
```
extract in a method to reuse in `set_destination_line`?

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,922 @@
+        location = self.env["stock.location"].browse(location_id)
+        move_line = self.env["stock.move.line"].browse(move_line_id)
+        if not location.exists() or not move_line.exists():
+            return self._response_for_start(message=self.msg_store.record_not_found())
```
This comment applies to several methods. I don't think the handling of no location and no package level/move line should be the same.

* No location: we are doomed anyway, doesn't make sense. Abort.
* No package level or no move line: for some reason, we a record, but we still have the rest of move lines / levels to move from the location, we should be able to continue with the next action. I think returning `start_or_recover()` should work

sebalix replied:
I agree, my first version was using `start_or_recover` :p and I removed that to simplify the checks as I said to myself "this case should never happened".

guewen commented on shopfloor/tests/test_location_content_transfer_set_destination_package_or_line.py:
```diff
@@ -0,0 +1,504 @@
+        #     message=self.service.msg_store.location_content_transfer_complete(
+        #         self.env["stock.location"]
+        #     ),
+        # )
```
still needed?

simahawk reviewed (commented):
Overall LG. Agree w/ @guewen 's comments

simahawk committed:
location transfer: improve pkg level parser

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,921 @@
+        ):
+            new_moves = self._create_moves_from_location(location)
+            new_moves._action_confirm(merge=False)
+            new_moves._action_assign()
```
seems like we can have an issue here, in case we can't assign (goods reserved already?), we won't be able to move the content or only partially, delete the moves and return an error message.

sebalix replied:
fixed by @simahawk in sebalix#1

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,921 @@
+        if (
+            not move_lines
+            and menu.allow_move_create
+            and len(menu.picking_type_ids) == 1
```
We must also check that the scanned location equal or a sublocation of the picking type default source location.

guewen commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -0,0 +1,921 @@
+        # then prepare new stock moves to move goods from the scanned location.
+        menu = self.work.menu
+        if (
+            not move_lines
```
Check that we have **no** move line at all in the source location, otherwise, if we create lines that can't be assigned, we have an infinite loop.

guewen replied:
And we have to add a search of stock.move draft or confirmed with same location_id, if we do, bail out with an error

simahawk committed:
location transfer bknd: add msg for completed item

simahawk committed:
location transfer bknd: always include source loc in data

simahawk committed:
location transfer bknd: rollback move lines and fail if not assigned

simahawk committed:
location transfer bknd: get rid of 'confirm_*' states

simahawk commented:
@sebalix pls check this when you get back https://github.com/sebalix/wms/pull/1

sebalix commented:
@simahawk sebalix#1 merged, unit tests completed and squashed in your own commits.
Regarding the draft package level which shouldn't be there once processed with `/stock_out_package`, I removed it manually (see https://github.com/camptocamp/wms/pull/27/commits/91251ba62974e32336d7c7a070ffc4bb8968b277 ).
If it's OK for you I'll squash the last fixups before the merge.

sebalix commented:
@simahawk done
